contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import { AdapterInterface } from "./interfaces/AdapterInterface.sol";
+
+/**
+ * @title ForwarderBase
+ * @notice This contract expects to receive messages and tokens from the Hub Pool on L1 and forwards them to targets on L3. Messages
+ * are intended to originate from the re-router adapter on L1 which will send messages here to re-route them to the corresponding L3
+ * spoke pool. It rejects messages which do not originate from the cross domain admin, which should be set to the hub pool.
+ * @dev Base contract designed to be deployed on L2 to re-route messages from L1 to L3. If a message is sent to this contract which
+ * came from the L1 cross domain admin, then it is routed to the L3 via the canonical messaging bridge. Tokens sent from L1 are stored
+ * temporarily on this contract. Any EOA can initiate a bridge of these tokens to the target `l3SpokePool`.
+ * @custom:security-contact bugs@across.to
+ */
+abstract contract ForwarderBase is UUPSUpgradeable, AdapterInterface {
+    /**
+     * @notice Struct containing all information needed in order to send a message to the next layer.
+     * @dev We need to know three different addresses:
+     *   - The address of the adapter to delegatecall, so that we may propagate message and token relays.
+     *   - The address of the next layer's forwarder, so that we can identify the intended recipient of message and token relays.
+     *   - The address of the next layer's spoke pool, so that we can determine whether the next chain contains the target spoke pool.
+     */
+    struct RouteIndex {
```

Comment:
Why call this an index?

Reply:
I'm trying to find the right word, but I am envisioning each of these contracts as a node in a graph/tree, and in order for each node to connect to subsequent layers, they need to know the info here, so I'm almost "indexing" in the graph to determine what that edge looks like.
... it's a stretch.

Reply:
hmm i like the visual of the graph/tree but I don't understand index. Naming is important, think on it some more.